cosign/verifier.go

```diff
@@ -0,0 +1,324 @@
+	}
+
+	return &Verifier{
+		// Set the name of the verifier from the provided options.
+		name: opts.Name,
+		// Set the keys map from the provided options.
+		truststore: NewWithOpts(opts),
+		// Set the cosign check options from the provided options.
+		CheckOpts: checkOpts,
+	}, nil
+}
+
+// NewCheckOpts updates the signature verifierOpts by verifierOptions.
+func NewCheckOpts(ctx context.Context, c *verify.VerifyCommand) (opts *cosign.CheckOpts, err error) {
+	// initialize the cosign check options
+	opts = &cosign.CheckOpts{}
+
+	if c.CheckClaims {
+		opts.ClaimVerifier = cosign.SimpleClaimVerifier
+	}
+
+	// If we are using signed timestamps, we need to load the TSA certificates
+	if c.TSACertChainPath != "" || c.UseSignedTimestamps {
+		tsaCertificates, err := cosign.GetTSACerts(ctx, c.TSACertChainPath, cosign.GetTufTargets)
+		if err != nil {
+			return nil, fmt.Errorf("unable to load TSA certificates: %w", err)
+		}
+		opts.TSACertificate = tsaCertificates.LeafCert
+		opts.TSARootCertificates = tsaCertificates.RootCert
+		opts.TSAIntermediateCertificates = tsaCertificates.IntermediateCerts
+	}
+
+	if !c.IgnoreTlog {
+		if c.RekorURL == "" {
+			c.RekorURL = defaultRekorURL
+		}
+
+		rekorClient, err := rekor.NewClient(c.RekorURL)
+		if err != nil {
+			return nil, fmt.Errorf("creating Rekor client: %w", err)
+		}
+		opts.RekorClient = rekorClient
+
+		// This performs an online fetch of the Rekor public keys, but this is needed
+		// for verifying tlog entries (both online and offline).
+		opts.RekorPubKeys, err = cosign.GetRekorPubs(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("getting Rekor public keys: %w", err)
+		}
+	}
+	return opts, nil
+}
+
+// Name returns the name of the verifier.
+func (v *Verifier) Name() string {
+	return v.name
+}
+
+// Type returns the type of the verifier which is always `cosign`.
+func (v *Verifier) Type() string {
+	return cosignVerifierType
+}
+
+// Verifiable returns true if the artifact is a cosign signature.
+func (v *Verifier) Verifiable(artifact ocispec.Descriptor) bool {
+	return artifact.ArtifactType == artifactTypeCosign &&
+		artifact.MediaType == ocispec.MediaTypeImageManifest
+}
+
+// Verify verifies the cosign signature.
+func (v *Verifier) Verify(ctx context.Context, opts *ratify.VerifyOptions) (*ratify.VerificationResult, error) {
+	v.MapSigVerifier(ctx, opts)
+
+	subjectRef, err := registry.ParseReference(opts.Subject)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse subject reference: %w", err)
+	}
+
+	signatureLayers, err := getSignatureBlobDesc(ctx, opts.Store, subjectRef, opts.SubjectDescriptor)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get signature blob descriptor: %w", err)
+	}
+
+	numResults := len(signatureLayers)
+	if numResults == 0 {
+		return nil, fmt.Errorf("unable to locate reference with artifactType %s", artifactTypeCosign)
+	}
+
+	signatureDesc := signatureLayers[numResults-1]
```

Comment:
why do we only keep the last layer?

Reply:
It feels like a cosign limit. We can talk about that offlinw